grails-app/services/org/pih/warehouse/inventory/CycleCountService.groovy

```diff
@@ -100,6 +100,66 @@ class CycleCountService {
         } as List<CycleCountCandidate>
     }
 
+    List<CycleCountCandidateInProgress> getCandidatesInProgress(CycleCountCandidateFilterCommand command, String facilityId) {
+        if (command.hasErrors()) {
+            throw new ValidationException("Invalid params", command.errors)
+        }
+        Integer max = command.format == "csv" ? null : command.max
+        Integer offset = command.format == "csv" ? null : command.offset
+        Location facility = Location.read(facilityId)
+        // Store added aliases to avoid duplicate alias exceptions for product
+        // This could happen when params.searchTerm and e.g. sort by product is applied
+        Set<String> usedAliases = new HashSet<>()
+        return CycleCountCandidateInProgress.createCriteria().list(max: max, offset: offset) {
+            eq("facility", facility)
+            if (command.searchTerm) {
+                createProductAlias(delegate, usedAliases)
+                or {
+                    ilike("product.productCode", "%${command.searchTerm}%")
+                    ilike("product.name", "%${command.searchTerm}%")
+                }
+            }
+            if (command.categories) {
+                usedAliases.add("product")
+                createAlias("product", "product", JoinType.INNER_JOIN)
+                "in"("product.category", command.categories)
+            }
+            if (command.internalLocations) {
+                or {
+                    command.internalLocations.each {
+                        ilike("internalLocations", "%${it}%")
+                    }
+                }
+            }
+            if (command.catalogs) {
+                createProductAlias(delegate, usedAliases)
+                createAlias("product.productCatalogItems", "productCatalogItems")
+                usedAliases.add("productCatalogItems")
+                "in"("productCatalogItems.productCatalog", command.catalogs)
+            }
+            if (command.tags) {
+                createProductAlias(delegate, usedAliases)
+                createAlias("product.tags", "tags")
+                usedAliases.add("tags")
+                "in"("tags.id", command.tags.collect { it.id })
+            }
+            if (command.abcClasses) {
+                "in"("abcClass", command.abcClasses)
+            }
+
+            if (command.statuses) {
+                inList("status", command.statuses)
+            }
+
+            if (command.negativeQuantity) {
+                gt("negativeItemCount", 0)
+            }
+
+            applySortOrder(command.sort, command.order, delegate, usedAliases)
+
+        } as List<CycleCountCandidateInProgress>
```

Comment:
I know the code is pretty much the same (except a few filters) as in the candidates endpoint, but I felt like it's "form over substance" to make those two service methods reusable with some if conditions. I feel like sometimes it's better to copy the code even if two places have to be refactored in the future, than to spend days thinking about how to make it reusable. Those are two easy list methods so I assume it's better to live with such version, but let me know your thoughts.

Reply:
I agree. Just because two things have essentially the same code, doesn't mean we should force them to use a common method. Sometimes they just do similar things and that's okay.

I think there's a way to look at it where CycleCountCandidate is an "enhanced" version of CycleCountCandidateInProgress, but yeah it feels a little forced.

If anything, I'd build a CycleCountXYZ interface that CycleCountCandidate and CycleCountCandidateInProgress can both extend from, which would contain all the common fields. But I do agree with you that it might be overkill for right now, especially since we want to make this change as simple as possible.

Maybe we can revisit it post-release.